### PR TITLE
One-Off Dynos

### PR DESCRIPTION
**Pull request by sbeck14**
Creating a formation with the `oneoff` flag set to true marks the formation as a "one-off". This instructs Akkeris to create a single pod and not have it restart when the container exits.

Provides ability to override app environment variables as well as the app image.

### CODE REVIEW

**Comment by sbeck14** [comment]
Needs improvements to the region-api (in progress, https://github.com/akkeris/region-api/pull/326) as well as the apps-watcher (to remove finished one-off pods and report to the controller)

File: package-lock.json [hunk 1]
```diff
@@ -1054,7 +1054,7 @@
     },
     "chalk": {
       "version": "1.1.3",
-      "resolved": "https://registry.npmjs.org/chalk/-/chalk-1.1.3.tgz",
+      "resolved": "http://registry.npmjs.org/chalk/-/chalk-1.1.3.tgz",
```
Comment (calaway) [inline package-lock.json:1057]: What's the deal with these changes? Is there a reason not to use `https`?

File: lib/formations.js [hunk 2]
```diff
@@ -197,6 +216,43 @@ function formation_payload_check(payload, sizes) {
     'A healthcheck was specified for a non-web based application, the path should not be set.');
   assert.ok(payload.type === 'web' || (payload.type !== 'web' && !payload.removeHealthcheck),
     'Cannot remove healthcheck on non web based dyno.');
+  assert.ok(payload.type === 'web' || payload.oneoff || (payload.type !== 'web' && payload.command !== '' && payload.command),
+    'A run command must be specified for any type outside of a web process.');
+  assert.ok(!payload.oneoff || payload.type !== 'web' || (payload.oneoff && payload.type !== 'web'),
```
Comment (trevorlinton) [inline lib/formations.js:221]: This seems wrong to me, what if someone creates a `worker` formation type? Wouldn't that get trapped here?

**Review by trevorlinton — CHANGES_REQUESTED** [review]
So, one thought, how will we handle billing on this? Creating a formation type implicitly creates a billing record until the formation is deleted.  Meaning, even after the one off dyno runs it will end up continously incurring charges unless (when it exists) it removes the record from the formation.

In addition, how will someone re-run a one off dyno, do they need to delete then recreate the formation of the one off dyno?

From a akkerisCI/build system stand point, all we actually need is the functions that are in `lib/alamo.js`, the rest of the code allows the end user to create a formation type of a one off dyno.  Perhaps we can split these two things out and revisit how we handle a user creating a one-off dyno as that might be a more complex to build out that then it is just to provide the AkkerisCI primitive for @calaway to use while building out AkkerisCI/Actions (and moving the buildshuttle code into the controller).  

Does that make sense?

**Comment by sbeck14** [comment]
> So, one thought, how will we handle billing on this? Creating a formation type implicitly creates a billing record until the formation is deleted.  Meaning, even after the one off dyno runs it will end up continously incurring charges unless (when it exists) it removes the record from the formation.

Will have to look into this one more- we could just have normal billing ignore formations with the `oneoff` flag and then handle it differently when actually triggering a one-off

> In addition, how will someone re-run a one off dyno, do they need to delete then recreate the formation of the one off dyno?
>
> From a akkerisCI/build system stand point, all we actually need is the functions that are in `lib/alamo.js`, the rest of the code allows the end user to create a formation type of a one off dyno.  Perhaps we can split these two things out and revisit how we handle a user creating a one-off dyno as that might be a more complex to build out that then it is just to provide the AkkerisCI primitive for @calaway to use while building out AkkerisCI/Actions (and moving the buildshuttle code into the controller).  

The formation is just the configuration for the one-off dyno, it doesn't actually run it. @calaway's code will have to handle triggering the function in `lib/alamo.js`. 

If we want to go a different route than reusing the formation primitive we can, I just thought that was the plan after our discussion.

File: package-lock.json (hunk 1 above)
Comment (sbeck14) [reply package-lock.json:1057]: 🤷🏻‍♂️ npm did this so I committed it. Can remove the change if you think it's best

File: lib/formations.js (hunk 2 above)
Comment (sbeck14) [reply lib/formations.js:221]: AFAIK the only actual special type of named dyno is "web". I don't think we do any checks in the code for "worker", and so it's basically "web" and "non-web" dynos. If someone created a one-off named "worker" then I don't think it should be treated any differently, only when they try and create a one-off "web" should we throw an error.

Thoughts?

Comment (trevorlinton) [reply lib/formations.js:221]: On second review this is accurate, i'm brain dead today I tells ya.

**Comment by trevorlinton** [comment]
ok, sorry, i’m not trying to be an obstruction but i’m wondering if we need to separate the concept of an internal call to create a one-off pod in kubernetes for tasks like actions or builds with the concept of a user requesting to run a single command via creating a one-off dyno.

For Calaways needs can't he just call https://github.com/akkeris/controller-api/pull/257/files#diff-2a698b7b055a1f06f523040f96b6e1fa415a0db4741df9a67ac298ed710fa21aR918 to create a one off pod for actions? and we can separate out the formation part of it and address those one off dyno concerns (billing, etc) separately?

E.g., what i'm saying is there's a subtle difference between a one-off dyno and one-off pod.  Calaway needs to create a one-off pod (in the way that i'm thinking about it) and a user needs to create a one-off dyno.  A user requesting a one-off dyno has some additional challenges that i'm wondering if it is easier for us to separate out this PR into two PRs (one for one-off pods, merge that then open a separate one for one-off dynos and address those billing concerns/etc). 

Hopefully this is a bit more clear, and i'll defer to you and calaway on the appropriate course of action if you feel its a non-issue and I'm just not clearly understanding this.

**Review by trevorlinton — APPROVED** [review]

**Comment by sbeck14** [comment]
Ready for review:
- `oneoff_options` has been renamed to `options` so that we can use it for other formations in the future
- Ability to stop one-off dynos has been added via `DELETE /apps/[app]/dynos/[dyno type]`
- One-off dynos have been excluded from all other restart operations
- Regex for image tag has been updated for codacy

The current codacy issue is not actually an issue, seems like their parser isn't the best.

**Comment by sbeck14** [comment]
Note - As part of akkeris actions, we need to implement handle billing differently for one-off dynos. 

This could be as simple as setting formation quantity to 0 when creating a one-off dyno, setting it to 1 when triggering it, and then setting it back to 0 when the apps-watcher reports that it has completed.

It could also be as complicated as excluding one-off formations from the standard invoices calculation and calculating it in a different way based on run history or some other metric.

Just something to keep in mind.

**Review by calaway — APPROVED** [review]

**Pull request merged**